pkg/chunk/purger/purger.go

```diff
@@ -0,0 +1,578 @@
+	}
+
+	dataPurger.runWorkers()
+	go dataPurger.jobScheduler()
+
+	err := dataPurger.loadInprocessDeleteRequests()
+	if err != nil {
+		return nil, err
+	}
+
+	return &dataPurger, nil
+}
+
+// Run starts workers, job scheduler and a loop for feeding delete requests to them for processing
+func (dp *DataPurger) Run() {
+	dp.loop()
+}
+
+// Stop stops all background workers/loops
+func (dp *DataPurger) Stop() {
+	close(dp.quit)
+	dp.wg.Wait()
+}
+
+func (dp *DataPurger) loop() {
+	dp.wg.Add(1)
+	defer dp.wg.Done()
+
+	pullDeleteRequestsToPlanDeletesTicker := time.NewTicker(time.Hour)
+	defer pullDeleteRequestsToPlanDeletesTicker.Stop()
+
+	for {
+		select {
+		case <-pullDeleteRequestsToPlanDeletesTicker.C:
+			err := dp.pullDeleteRequestsToPlanDeletes()
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "error pulling delete requests for building plans", "err", err)
+			}
+		case <-dp.quit:
+			return
+		}
+	}
+}
+
+// we send all the delete plans to workerJobChan and then start checking for status on workerJobExecutionStatusChan
+func (dp *DataPurger) jobScheduler() {
+
+	pendingPlansCount := make(map[string]int)
+	pendingPlansCountMtx := sync.Mutex{}
+
+	dp.wg.Add(1)
+
+	go func() {
+
+		for {
+			select {
+			case jobExecutionStatus := <-dp.workerJobExecutionStatusChan:
+				if jobExecutionStatus.err != nil {
+					level.Error(jobExecutionStatus.logger).Log("msg", "error executing delete plan",
+						"plan_no", jobExecutionStatus.planNo, "err", jobExecutionStatus.err)
+				} else {
+					err := dp.removeDeletePlan(context.Background(), jobExecutionStatus.userID, jobExecutionStatus.deleteRequestID, jobExecutionStatus.planNo)
+					if err != nil {
+						level.Error(jobExecutionStatus.logger).Log("msg", "error removing delete plan",
+							"plan_no", jobExecutionStatus.planNo, "err", err)
+					} else {
+						pendingPlansCountMtx.Lock()
+						pendingPlansCount[jobExecutionStatus.deleteRequestID]--
+
+						if pendingPlansCount[jobExecutionStatus.deleteRequestID] == 0 {
+							level.Info(jobExecutionStatus.logger).Log("msg", "finished execution of all plans, cleaning up and updating status of request")
+
+							err := dp.deleteStore.UpdateStatus(context.Background(), jobExecutionStatus.userID, jobExecutionStatus.deleteRequestID, chunk.Processed)
+							if err != nil {
+								level.Error(jobExecutionStatus.logger).Log("msg", "error updating delete request status to process", "err", err)
+							}
+
+							delete(pendingPlansCount, jobExecutionStatus.deleteRequestID)
+							pendingPlansCountMtx.Unlock()
+
+							dp.inProcessRequestIDsMtx.Lock()
+							delete(dp.inProcessRequestIDs, jobExecutionStatus.deleteRequestID)
+							dp.inProcessRequestIDsMtx.Unlock()
+						} else {
+							pendingPlansCountMtx.Unlock()
+
+						}
+
+					}
+				}
+			case <-dp.quit:
+				dp.wg.Done()
+				return
+			}
+		}
+	}()
+
+	dp.wg.Add(1)
+	defer dp.wg.Done()
+
+	for {
+		select {
+		case requestWithLogger := <-dp.executePlansChan:
```

Comment:
```suggestion
		case req := <-dp.executePlansChan:
```

No need to call it requestWithLogger.